File: src/java/magmac/app/compile/rule/StripRule.java
package magmac.app.compile.rule;

import magmac.app.compile.error.CompileResult;
import magmac.app.compile.node.Node;

/**
 * Rule that trims input before lexing and appends optional strings when
 * generating output.  The strings are read from {@code before} and
 * {@code after} keys on the node and default to empty if absent.
 */
public record StripRule(Rule rule) implements Rule {
    @Override
    public CompileResult<Node> lex(String input) {
        return this.rule.lex(input.strip());
    }

    @Override
    public CompileResult<String> generate(Node node) {
        var before = node.findString("before").orElse("");
        var after = node.findString("after").orElse("");
        return this.rule.generate(node).mapValue(result -> before + result + after);
    }
}

File: src/web/magmac/app/compile/rule/StripRule.ts
export class StripRule {
	 StripRule( rule : Rule) : public {this( "", rule, "");;}
	public lex( input : String) : CompileResult<Node> {return this.rule.lex( input.strip( ));;}
	public generate( node : Node) : CompileResult<String> { let before : var=node.findString( this.beforeKey).orElse( ""); let after : var=node.findString( this.afterKey).orElse( "");return this.rule.generate( node).mapValue( 0);;}
}
